runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -290,12 +333,83 @@ private void setLastError(String error) {
     lastErrorTime.set(DateTime.now());
   }
 
+  protected abstract void shutdownInternal();
+
   public static class WindmillStreamShutdownException extends RuntimeException {
     public WindmillStreamShutdownException(String message) {
       super(message);
     }
   }
 
+  /**
+   * Request observer that allows resetting its internal delegate using the given {@link
+   * #requestObserverSupplier}.
+   */
+  @ThreadSafe
+  private static class ResettableRequestObserver<RequestT> implements StreamObserver<RequestT> {
+    private final Supplier<StreamObserver<RequestT>> requestObserverSupplier;
+
+    @GuardedBy("delegateRequestObserver")
+    private final AtomicReference<StreamObserver<RequestT>> delegateRequestObserver;
+
+    @GuardedBy("delegateRequestObserver")
+    /* Indicates if onCompleted() has been called for the current delegateRequestObserver instance.
+    Reset to false when reset() is called. */
+    private volatile boolean isClosed;
```

Comment:
reverted back to streamClosed then
i was trying to find a way to handle that logic/signal closer to the source, but if it introduces racy behavior, better to avoid